pow/src/main/java/tech/pegasys/teku/pow/ValidatingEth1EventsPublisher.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.pow;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.Optional;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.pow.api.Eth1EventsChannel;
+import tech.pegasys.teku.pow.event.DepositsFromBlockEvent;
+import tech.pegasys.teku.pow.exception.InvalidDepositEventsException;
+
+public class ValidatingEth1EventsPublisher extends DelegatingEth1EventsChannel {
+  private Optional<UInt64> lastPublishedDeposit = Optional.empty();
+
+  public ValidatingEth1EventsPublisher(final Eth1EventsChannel delegate) {
+    super(delegate);
+  }
+
+  public synchronized void setLastestPublishedDeposit(final UInt64 latestPublishedDeposit) {
+    checkNotNull(latestPublishedDeposit);
+    if (!lastPublishedDeposit.isEmpty()) {
+      throw new IllegalStateException("Latest published deposit is already set");
+    }
+    this.lastPublishedDeposit = Optional.of(latestPublishedDeposit);
+  }
+
+  @Override
+  public synchronized void onDepositsFromBlock(final DepositsFromBlockEvent event) {
+    assertDepositEventIsValid(event);
+    lastPublishedDeposit = Optional.of(event.getLastDepositIndex());
+
+    delegate.onDepositsFromBlock(event);
+  }
+
+  private void assertDepositEventIsValid(final DepositsFromBlockEvent event) {
+    final Optional<UInt64> expectedFirstDepositIndex = lastPublishedDeposit.map(UInt64::increment);
+    final boolean depositIndexIsValid =
+        expectedFirstDepositIndex
+            .map(expected -> expected.equals(event.getFirstDepositIndex()))
+            .orElse(true);
```

Comment:
I suspect this misses the case where we miss the very first deposit.  We won't have replayed anything from disk so the last deposit index will be unset and we should then expect the first deposit to have index 0.

Reply:
good call - fixed